File: algorithms/0018-4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        r = list()
        for a in range(len(nums)):
            if a > 0 and nums[a] == nums[a - 1]:
                continue
            for b in range(a + 1, len(nums)):
                if b > a + 1 and nums[b] == nums[b - 1]:
                    continue
                c, d = b + 1, len(nums) - 1
                while c < d:
                    if c > b + 1 and d < len(nums) - 1 and nums[c] == nums[c - 1] and nums[d] == nums[d + 1]:
                        c += 1
                        d -= 1
                        continue
                    if nums[a] + nums[b] + nums[c] + nums[d] < target:
                        c += 1
                    elif nums[a] + nums[b] + nums[c] + nums[d] > target:
                        d -= 1
                    else:
                        r.append([nums[a], nums[b], nums[c], nums[d]])
                        c += 1
                        d -= 1
        return r

File: algorithms/0018-4sum.go
package algorithms

func fourSum(nums []int, target int) [][]int {
	sort.Ints(nums)
	var result [][]int
	for a := 0; a < len(nums) - 3; a++ {
		if a > 0 && nums[a] == nums[a - 1] {
			continue
		}
		for b := a + 1; b < len(nums) - 2; b++ {
			if b > a + 1 && nums[b] == nums[b - 1] {
				continue
			}
			c, d := b + 1, len(nums) - 1
			for c < d {
				if c > b + 1 && d < len(nums) - 1 && nums[c] == nums[c - 1] && nums[d] == nums[d + 1] {
					c++
					d--
					continue
				}
				if nums[a] + nums[b] + nums[c] + nums[d] < target {
					c++
				} else if nums[a] + nums[b] + nums[c] + nums[d] > target {
					d--
				} else {
					result = append(result, []int{nums[a], nums[b], nums[c], nums[d]})
					c++
					d--
				}
			}
		}
	}
	return result
}
